### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): Classification of `Spec R ⟶ X` with `R` local.

### PR DESCRIPTION
**Pull request by erdOne**
---

- [x] depends on: #15259

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:95]:
```suggestion
  rw [← hV.fromSpecStalk_eq_fromSpecStalk hxV, IsAffineOpen.fromSpecStalk, Scheme.comp_app,
    hV.fromSpec_app_of_le _ hVU, ← X.presheaf.germ_res (homOfLE hVU) x hxV]
  simp [Category.assoc, ← ΓSpecIso_inv_naturality_assoc]
```

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:179]:
```suggestion
lemma Spec_map_stalkClosedPointIso :
```

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Stalk.lean:179]: Or maybe `Spec.map_stalkClosedPointIso`?

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:203]:
```suggestion
Given a local ring `(R, 𝔪)` and a morphism `f : Spec R ⟶ X`,
they induce a ring homomorphim `φ : 𝒪_{X, f 𝔪} ⟶ R`.
```

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:216]: Do you understand why `infer_instance` does not work?

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:197]: Also this could go to ~`AlgebraicGeometry.Spec`~ `AlgebraicGeometry.Scheme`.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:309]:
```suggestion
@[simps]
noncomputable
def SpecToEquivOfLocalRing :
```
seems to produce reasonable lemmas here.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Stalk.lean:179]: There are more occurrences below. I think we should decide whether `Spec_map_`, `Spec.map_` or `Spec_` is the convention.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:75]:
```suggestion
  rw [IsAffineOpen.fromSpecStalk, Scheme.comp_val_base_apply]
  rw [← hU.primeIdealOf_eq_map_closedPoint ⟨x, hxU⟩, hU.fromSpec_primeIdealOf ⟨x, hxU⟩]
```

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:232]: We can get rid of this `erw` by adding a lemma
```
lemma preimage_top (f : X ⟶ Y) : f ⁻¹ᵁ ⊤ = ⊤ :=
  rfl
```

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:259]: With the lemma from above:
```suggestion
  simp_rw [← preimage_top (Spec.map f)]
  rw [← (Spec.map f).app_eq_appLE]
```

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:216]: I hope/believe #17403 will solve this.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:232]: It seems to throw a motive incorrect error for me.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:179]: Probably `Spec_map_` I guess?

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Stalk.lean:232]: This
```
lemma germ_stalkClosedPointTo_Spec {R S : CommRingCat} [LocalRing S] (φ : R ⟶ S):
    (Spec R).presheaf.germ ⊤ _ trivial ≫ stalkClosedPointTo (Spec.map φ) =
      (ΓSpecIso R).hom ≫ φ := by
  rw [stalkClosedPointTo, Scheme.stalkMap_germ_assoc, ← Iso.inv_comp_eq,
    ← ΓSpecIso_inv_naturality_assoc]
  simp_rw [preimage_top]
  rw [germ_stalkClosedPointIso_hom]
  rw [Iso.inv_hom_id, Category.comp_id]
```
works for me.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:133]: Cool!

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:192]: Nice catch!

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:232]: Oh `simp_rw`. Makes sense.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:95]: Hmm this doesn't work for me.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:15]: I think it should be `specializes`?

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Stalk.lean:15]: I am not a native speaker, so you might very well be right. I thought that "the `y`" is plural here.

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean:15]: Oh yes you're right the `y` should be plural

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Stalk.lean]: This is strange, I just copy pasted this into your new version and it still works for me. Probably it is not very important, I just like that we can get rid of the terminal `rfl`.

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:242]:
```suggestion
  have : (Spec.map f ≫ X.fromSpecStalk x).base (closedPoint R) = x := by
```

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Stalk.lean:303]:
```suggestion
  toFun f := ⟨f.base (closedPoint R), Scheme.stalkClosedPointTo f, inferInstance⟩
```

File: Mathlib/AlgebraicGeometry/Stalk.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Stalk.lean]: Oh I thought it was `simp only`. Thanks!

**Review by chrisflav — APPROVED** [review]
Thanks, LGTM!

**Comment by kbuzzard** [comment]
Many thanks for all this work!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
